### PR TITLE
Eclair RPC 0.4.1

### PR DESCRIPTION
**Pull request by rorp**

### CODE REVIEW

File: app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala [hunk 1]
```diff
@@ -8,6 +8,12 @@ import java.util.UUID
 import org.bitcoins.commons.jsonmodels._
 import org.bitcoins.commons.jsonmodels.bitcoind.RpcOpts.LabelPurpose
 import org.bitcoins.commons.jsonmodels.bitcoind._
+import org.bitcoins.commons.jsonmodels.eclair.ChannelCommandState.{
```
Comment (benthecarman) [inline app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala]: nit: `ChannelCommandState._`

File: app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala [hunk 2]
```diff
@@ -257,6 +254,18 @@ object JsonReaders {
...
 
+  implicit object BitcoinNetworkReads extends Reads[BitcoinNetwork] {
+
+    override def reads(json: JsValue): JsResult[BitcoinNetwork] =
+      SerializerUtil.processJsString {
+        case "mainnet"              => MainNet
+        case "testnet" | "testnet3" => TestNet3
+        case "regtest"              => RegTest
+        case err @ _ =>
+          throw new RuntimeException(s"Unknown Bitcoin network `$err`")
```
Comment (benthecarman) [inline app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala]: Should use `BitcoinNetworks.fromString`

File: app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala [hunk 3]
```diff
@@ -843,6 +889,35 @@ object JsonReaders {
...
   }
 
+  implicit val channelCommandStateReads: Reads[ChannelCommandState] = Reads {
+    jsValue =>
+      SerializerUtil.processJsString { s =>
+        if (s == "ok") {
+          OK
+        } else if (s.startsWith("created channel ")) {
+          ChannelOpened
+        } else if (s.startsWith("closed channel ")) {
```
Comment (benthecarman) [inline app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala]: Are these supposed end with a `" "`?

Comment (benthecarman) [reply app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala]: Also, would be nice to just have a `ChannelCommandState.fromString`

File: app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala [hunk 4]
```diff
@@ -1129,6 +1204,19 @@ object JsonReaders {
...
+  implicit val channelStatsDirectionReads: Reads[ChannelStats.Direction] =
+    Reads { json =>
+      SerializerUtil.processJsString { s =>
+        if (s.toUpperCase == "IN") {
+          ChannelStats.In
+        } else if (s.toUpperCase == "OUT") {
+          ChannelStats.Out
+        } else {
+          throw new RuntimeException(s"Unknown payment direction: `$s`")
+        }
```
Comment (benthecarman) [inline app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala]: Would be nice to just have a `ChannelStats.Direction.fromString`

File: app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala [hunk 5]
```diff
@@ -82,14 +102,23 @@ case class OpenChannelInfo(
     state: ChannelState.NORMAL.type
 ) extends ChannelInfo
 
+case class ActivatedFeature(feature: Feature, support: FeatureSupport)
+
+case class UnknownFeature(bitIndex: Int)
```
Comment (benthecarman) [inline app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala:117]: Should we add an invariant enforcing even/odd-ness?

File: app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala [hunk 6]
```diff
@@ -82,14 +102,23 @@ case class OpenChannelInfo(
...
     signature: ECDigitalSignature,
-    features: String,
+    features: Features,
     timestamp: Instant,
     nodeId: NodeId,
     rgbColor: String,
     alias: String,
-    addresses: Vector[InetSocketAddress])
+    addresses: Vector[InetSocketAddress],
+    unknownFields: String)
```
Comment (benthecarman) [inline app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala:131]: Isn't `unknownFields` in `features`?

Comment (rorp) [reply app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala:131]: No, see https://acinq.github.io/eclair/#nodes

File: app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala (hunk 3 above)
Comment (rorp) [reply app-commons/src/main/scala/org/bitcoins/commons/serializers/JsonReaders.scala]: > Are these supposed end with a `" "`?

Yes

File: app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala (hunk 5 above)
Comment (rorp) [reply app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala:117]: I believe it's out of the scope of this PR.

File: app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala [hunk 7]
```diff
@@ -28,18 +29,47 @@ import scala.concurrent.duration.FiniteDuration
...
+  def fromString(s: String): State =
+    if (s == "ok") {
+      ChannelCommandResult.OK
+    } else if (s.startsWith("created channel ")) {
+      ChannelCommandResult.ChannelOpened
+    } else if (s.startsWith("closed channel ")) {
+      ChannelCommandResult.ChannelClosed
+    } else {
+      ChannelCommandResult.Error(s)
+    }
```
Comment (benthecarman) [inline app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala:70]: last nit: should probably ignore case in these comparisons

Comment (rorp) [reply app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/eclair/EclairModels.scala:70]: There's no need in this: Eclair doesn't send (and log) messages in upper case.

**Pull request merged**